update authz concept page for the new v1beta1 policy (#5165)
istio/istio.io: content/en/docs/concepts/security/index.md

## Authorization

> 

Author: Done.

Istio's authorization feature provides mesh-level, namespace-level, and workload-level
access control on workloads in an Istio Mesh. It features:

- **Single API**, which is easy to use and maintain.

> - **Flexible semantic**, extra conditions are supported for finer grained access control.

Reviewer: "extra" and "finer grained" are compared to what?

Author: "extra" is compared to the first-class fields provided in the API, "finer grained" comes from the rich semantics in the conditions, like the support of JWT claims, etc.

Author: Done.

## Authorization

Istio's authorization feature provides mesh-level, namespace-level, and workload-level

> access control on workloads in an Istio Mesh. It features:

Reviewer 2: access control on workloads in an Istio Mesh. It ~features~ provides:

- **Workload-to-workload and end-user-to-workload authorization**.

> - **Simple API**, it includes a single [`AuthorizationPolicy` CRD](/docs/reference/config/authorization/authorization-policy/), which is easy to use and maintain.

Reviewer 2: **A Simple API**

> - **Flexible semantic**, operators can define custom conditions on Istio attributes.

Reviewer 2: **Flexible semantics**

    >}}

The above diagram shows the basic Istio authorization architecture. Operators
specify Istio authorization policies using `.yaml` files.

> Galley watches for changes to Istio authorization policies. It fetches the

Reviewer: I think we should remove this paragraph, or move it to the architecture document. In general, we want to be talking about the features and not how the features are implemented.

@rcaballeromx

Reviewer 3: @geeknoid I agree. This information doesn't belong here and we should move it to, at the very least an architecture section, or ideally the Architecture page under reference.

updated authorization policies if it sees any changes and pushes to Pilot. Pilot

> converts the authorization policies to Envoy filter and distributes the filter

Author: Sure, we can revisit this page later.

    >}}

The above diagram shows the basic Istio authorization architecture. Operators
specify Istio authorization policies using `.yaml` files.

> Galley watches for changes to Istio authorization policies and fetches the

Reviewer: I would remove this entire paragraph. In general, we want to avoid architectural discussions in the docs and focus on features.

Author: Removed.

updated authorization policies if it sees changes and pushes to Pilot. Pilot
converts the authorization policies to Envoy filter configuration and distributes the filter
to the Envoy proxies that are co-located with the workloads.

Each Envoy proxy runs an authorization engine that authorizes requests at
runtime. When a request comes to the proxy, the authorization engine evaluates
the request context against the current authorization policies, and returns the
authorization result, `ALLOW` or `DENY`.

> ### Implicit Enablement

Reviewer: ```suggestion
### Implicit enablement
```

There is no need to explicitly enable Istio's authorization feature, you just apply
the `AuthorizationPolicy` on **workloads** to enforce access control.

If no `AuthorizationPolicy` applies to a workload, no access control will be enforced,
In other words, all requests will be allowed.

If any `AuthorizationPolicy` applies to a workload, access to that workload is
denied by default, unless explicitly allowed by a rule declared in the policy.

> Currently `AuthorizationPolicy` only supports "ALLOW" action. This means that if

Reviewer: ```suggestion
Currently `AuthorizationPolicy` only supports `ALLOW` action. This means that if
```

multiple authorization policies apply to the same workload, the effect is additive.

### Authorization policy

> To configure an Istio authorization policy, you specify the [`AuthorizationPolicy` configuration object](

Reviewer: ```suggestion
To configure an Istio authorization policy, you create an [`AuthorizationPolicy` resource](
```

> /docs/reference/config/authorization/authorization-policy/). Like other Istio configuration objects, it

Reviewer: No need for the sentence about CRDs, not relevant to customers using Istio.

is defined as a Kubernetes `CustomResourceDefinition` [(CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
object.

> An `AuthorizationPolicy` includes a `selector` and a list of `rules`. The `selector`

Reviewer: ```suggestion
An authorization policy includes a selector and a list of rules. The selector
```

> specifies the **target** that the policy applies to. The `rules` specify **who**

Reviewer: ```suggestion
specifies the **target** that the policy applies to, while the rules specify **who**
```

root namespace, the policy applies to all namespaces in a mesh. The value of
root namespace is configurable, and the default is `istio-system`. If set to a
normal namespace, the policy will only apply to the specified namespace.

A workload `selector` can be used to further restrict where a policy applies.
The `selector` uses pod labels to select the target workload. The workload
selector contains a list of `{key: value}` pairs, where the `key` is the name of the label.
If not set, the authorization policy will be applied to all workloads in the same namespace
as the authorization policy.

The following example policy `allow-read` allows `"GET"` and `"HEAD"` access to
the workload with label `app: products` in the `default` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-read
  namespace: default
spec:
  selector:
    matchLabels:
      app: products
  rules:
  - to:
    - operation:
         methods: ["GET", "HEAD"]
{{< /text >}}

#### Value matching

Exact match, prefix match, suffix match, and presence match are supported for most

> of the field with a few exceptions (e.g., the `key` field under `when` section,

Reviewer: ```suggestion
of the field with a few exceptions (e.g., the `key` field under the `when` section,
```

> the `ipBlocks` under `source` section and the `ports` field under `to` section only support exact match).

Reviewer: ```suggestion
the `ipBlocks` under the `source` section and the `ports` field under the `to` section only support exact match).
```

- **Exact match**. i.e., exact string match.
- **Prefix match**. A string with an ending `"*"`. For example, `"test.abc.*"` matches `"test.abc.com"`, `"test.abc.com.cn"`, `"test.abc.org"`, etc.
- **Suffix match**. A string with a starting `"*"`. For example, `"*.abc.com"` matches `"eng.abc.com"`, `"test.eng.abc.com"`, etc.
- **Presence match**. `*` is used to specify anything but not empty. You can specify a field must be present using the format `fieldname: ["*"]`.
This means that the field can match any value, but it cannot be empty. Note that it is different from leaving a field unspecified, which means anything including empty.

> The following example policy allows allows access at paths with prefix `"/test/"` or suffix `"/info"`.

Reviewer: ```suggestion
The following example policy allows access at paths with prefix `"/test/"` or suffix `"/info"`.
```

apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: tester
  namespace: default
spec:
  selector:
    matchLabels:
      app: products
  rules:
  - to:
    - operation:
        paths: ["/test/*", "*/info"]
{{< /text >}}

#### Allow-all and deny-all

The example below shows a simple policy `allow-all` which allows full access to all
workloads in the `default` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-all
  namespace: default
spec:
  rules:
  - {}
{{< /text >}}

The example below shows a simple policy `deny-all` which denies access to all workloads
in the `admin` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: deny-all
  namespace: admin
spec:
  {}
{{< /text >}}

#### Custom conditions

You can also use the `when` section to specify additional conditions. For example, the following
`AuthorizationPolicy` definition includes a condition that `request.headers[version]` is either `"v1"` or `"v2"`.
In this case, the key is `request.headers[version]`, which is an entry in the Istio attribute `request.headers`,

> which is `map` data type.

Reviewer: ```suggestion
which is a map.
```

    >}}

The above diagram shows the basic Istio authorization architecture. Operators
specify Istio authorization policies using `.yaml` files.

Galley watches for changes to Istio authorization policies. It fetches the
updated authorization policies if it sees any changes and pushes to Pilot. Pilot
converts the authorization policies to Envoy filter configuration and distributes the filter
to the Envoy proxies that are co-located with the workloads.
Pilot watches for changes to Istio authorization policies. It fetches the

> updated authorization policies if it sees any changes. Pilot distributes Istio

Reviewer 2: Pilot watches for changes to Istio authorization policies~.~ and ~It~ fetches ~the~
updated authorization policies if it sees ~any~ changes.

### Enabling authorization

You enable Istio Authorization on **workloads** by applying the `AuthorizationPolicy`
custom resource that selects the workloads. Currently `AuthorizationPolicy` only
supports "ALLOW" action. This means that if multiple authorization policies apply
to the same workload, the effect is additive.

The following authorization policy enables access control on workloads with labels
`app: httpbin` and `version: v1` in namespace foo.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: httpbin
 namespace: foo
spec:
 selector:
   matchLabels:
     app: httpbin
     version: v1
 rules:
 - from:
   - source:
       principals: ["cluster.local/ns/default/sa/sleep"]
   to:
   - operation:
       methods: ["GET"]
   when:
   - key: request.auth.claims[iss]
     values: ["https://accounts.google.com"]
{{< /text >}}

### Authorization policy

> To configure an Istio authorization policy, you specify the [Authorization Policy](https://github.com/istio/api/blob/bfa91e88abf1af16d248917ebd3ab62036f7ca12/security/v1beta1/authorization.proto)

Reviewer: Actually, please change it to {{< source_branch_name >}}. This will track the right branch over time. We never want to point our docs to master.

> To configure an Istio authorization policy, you specify the [Authorization Policy](https://github.com/istio/api/blob/bfa91e88abf1af16d248917ebd3ab62036f7ca12/security/v1beta1/authorization.proto)

Author: I updated to link to the ref page: `/docs/reference/config/authorization/authorization-policy/` instead of using the API repo.

configuration object. Like other Istio configuration objects, it is defined as Kubernetes `CustomResourceDefinition` [(CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
object.

An `AuthorizationPolicy` includes a `selector` and a list of `rules`, the `selector` specifies

> **where** to apply the policy and the `rules` specify **who** is allowed to do **what** under which **conditions**.

Reviewer 4: specifies **where** to apply the policy => specifies the **target" that the policy applies to.

Specifically:

> - **where** refers to the `selector` section in the `AuthorizationPolicy`.

Reviewer 4: **where** => **target**

- **who** refers to the `from` section in the `rule` of the `AuthorizationPolicy`.
- **what** refers to the `to` section in the `rule` of the `AuthorizationPolicy`.
- **conditions** refers to the `when` section in the `rule` of the `AuthorizationPolicy`.

Workload selector decides where to apply the authorization policy. If not set,
the authorization policy will be applied to all workloads in the same namespace
as the authorization policy.

> 

Reviewer 4: We also need to mention "root namespace". This is how a mesh level policy is defined. My preference is to have a dedicated section talking about "policy scope and workload selector" (like we have in the [design doc](https://docs.google.com/document/d/1diwa9oYVwmLtarXb-kfWPq-Ul3O2Ic3b8Y21Z3R8DzQ/edit#heading=h.wq9jh71vsoq0)).

Author: Added the section `Scope and workload selector` for this.

Each rule has the following standard fields:

> - **`from`**: A list of sources for specifying the identities.

Reviewer 4: **`from`**: A list of `sources`.

Source does not only include identities. It can also be IPs, namespaces, etc.

> - **`to`**: A list of operations for specifying the operations.

Reviewer 4: A list of `operations`.

> - **`when`**: A list of conditions for specifying extra conditions.

Reviewer 4: **when**: Custom conditions.

An `AuthorizationPolicy` specification only applies to the namespace specified in the

> `metadata` section. All fields in the rule are optional. If you do not specify a field

Reviewer 5: Are you talking about a policy is applied to all instances in the namespace? If so, I think the "All fields in the rule are optional" is confusing here.

Author: I was trying to say all fields in the rule could be left unset, could you clarify what is confusing here?

Reviewer 5: It's confusing me because in the same paragraph you talk about policy scoping ("...only applies to the namespace specified...") and in the middle you talk about all fields are optional (which has nothing to do with policy scoping?), then ends the paragraph with "Istio applies the field to all instances", which seems like the policy scoping again.

Author: This part is removed now.

> or if you set its value to `*`, Istio applies the field to all instances.

Reviewer 4: There is a difference between leaving a field unset vs. set a field to `*`. 

* If you leave a field unset, there is no restriction on the field (can be empty). 
* If you set a field to "*", the field has to be non-empty. 

We need a section about exact match, prefix match, suffix match, presence match. See the [section](https://docs.google.com/document/d/1diwa9oYVwmLtarXb-kfWPq-Ul3O2Ic3b8Y21Z3R8DzQ/edit#bookmark=id.j08yqyvjd0nn) in design doc.

Author: Added a new section `Value matching` for this.

The example below shows a simple policy `allow-all` which allows full access to all
workloads in the `default` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-all
  namespace: default
spec:
  rules:
  - {}
{{< /text >}}

Here is another policy `allow-read` which allows `"GET"` and `"HEAD"` access to
the workload with label `app: products` in the `default` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-read
  namespace: default
spec:
  selector:
    matchLabels:
      app: products
  rules:
  - to:
    - operation:
         methods: ["GET", "HEAD"]
{{< /text >}}

In addition, we support prefix matching and suffix matching for all the fields
in a rule. For example, the following policy allows allows access at paths with
prefix `"/test/"` or suffix `"/info"`.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: tester
  namespace: default
spec:
  selector:
    matchLabels:
      app: products
  rules:
  - to:
    - operation:
        paths: ["/test/*", "*/info"]
{{< /text >}}

> You can also use the `when` section to specify additional constraints, for example, the following

Reviewer 4: constraints => conditions

> `AuthorizationPolicy` definition includes a constraint that `request.headers[version]` is either `"v1"` or `"v2"`.

Reviewer 4: constraint => condition.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: httpbin
 namespace: foo
spec:
 selector:
   matchLabels:
     app: httpbin
     version: v1
 rules:
 - from:
   - source:
       principals: ["cluster.local/ns/default/sa/sleep"]
   to:
   - operation:
       methods: ["GET"]
   when:
   - key: request.headers[version]
     values: ["v1", "v2"]
{{< /text >}}

In the case that the attribute is a `map`, for example `request.headers`, the
`key` is an entry in the map, for example `request.headers[version]`. The supported `key` values

> of a constraint are listed in the [conditions page](/docs/reference/config/authorization/conditions/).

Reviewer 4: constaint => condition

You can specify flexible source identities in the authorization policy. The following example
shows an `AuthorizationPolicy` that allows two sources (service account `cluster.local/ns/default/sa/sleep`
and namespace `dev`) to access the httpbin workload:

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: httpbin
 namespace: foo
spec:
 selector:
   matchLabels:
     app: httpbin
     version: v1
 rules:
 - from:
   - source:
       principals: ["cluster.local/ns/default/sa/sleep"]
   - source:
       namespaces: ["dev"]
   to:
   - operation:
       methods: ["GET", "POST"]
{{< /text >}}

In case you want to make a workload publicly accessible, you can set the
`principals` to `principals: ["*"]`. This value allows sources from **all (both authenticated and

> unauthenticated)** users and services, for example:

Reviewer 4: This is not right. If you set `principals: ["*"]`, it means principals field cannot be empty (authenticated users only). 

If you want to allow both authenticated and unauthenticated users, you need to leave "source" section empty.

Author: Updated the doc, we need to double check the implementation is doing this way.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: httpbin
 namespace: foo
spec:
 selector:
   matchLabels:
     app: httpbin
     version: v1
 rules:
 - from:
   - source:
       principals: ["*"]
   to:
   - operation:
       methods: ["GET", "POST"]
{{< /text >}}

To allow only **authenticated** users, use `source.principal: "*"` instead, for example:

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: httpbin
 namespace: foo
spec:
 selector:
   matchLabels:
     app: httpbin
     version: v1
 rules:
 - from:
   - source:

>        principals: ["cluster.local/ns/default/sa/sleep"]

Reviewer 4: Why do we need condition here? It should simply be:
```
- from:
  - source:
      principals: ["*"]
 to:
 - operation:
      methods: ["GET", "POST"]
```

   to:
   - operation:
       methods: ["GET", "POST"]
   when:
   - key: source.principal
     values: ["*"]
{{< /text >}}

### Using Istio authorization on plain TCP protocols

Istio authorization supports workloads using any plain TCP protocols, such as MongoDB. In this case,
you configure the authorization policy in the same way you did for the HTTP workloads.
The difference is that certain fields and conditions are only applicable to HTTP workloads.
These fields include:

> - The `request_principals` fields in the source section of the authorization policy object

Reviewer 4: fields => field

- The `hosts`, `methods` and `paths` fields in the operation section of the authorization policy object

> - The supported conditions are listed in the [constraints page](/docs/reference/config/authorization/conditions/).

Reviewer 4: constraints page => conditions page

root namespace, the policy applies to all namespaces in a mesh. The value of
root namespace is configurable, and the default is `istio-config`. If set to a
normal namespace, the policy will only apply to the specified namespace.

A workload `selector` can be used to further restrict where a policy applies.
The `selector` uses pod labels to select the target workload. The workload
selector contains a list of `{key: value}` pairs, where the `key` is the name of the label.
If not set, the authorization policy will be applied to all workloads in the same namespace
as the authorization policy.

The following example policy `allow-read` which allows `"GET"` and `"HEAD"` access to
the workload with label `app: products` in the `default` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-read
  namespace: default
spec:
  selector:
    matchLabels:
      app: products
  rules:
  - to:
    - operation:
         methods: ["GET", "HEAD"]
{{< /text >}}

#### Value matching

Exact match, prefix match, suffix match, and presence match are supported for most

> of the field with a few exceptions (e.g., "when/key" only supports exact match).

Author: I think `ip` and `port` are also not supported

> of the field with a few exceptions (e.g., "when/key" only supports exact match).

Reviewer 4: Let's clearly call out these fields only support exact match, either in this doc, or document it in API reference. So that people do not need to guess.

This means that the field can match any value, but it cannot be empty. Note that it is different from leaving a field unspecified, which means anything including empty.

The following example policy allows allows access at paths with prefix `"/test/"` or suffix `"/info"`.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: tester
  namespace: default
spec:
  selector:
    matchLabels:
      app: products
  rules:
  - to:
    - operation:
        paths: ["/test/*", "*/info"]
{{< /text >}}

#### Allow-all and deny-all

The example below shows a simple policy `allow-all` which allows full access to all
workloads in the `default` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-all
  namespace: default
spec:
  rules:
  - {}
{{< /text >}}

The example below shows a simple policy `deny-all` which denies access to all workloads
in the `admin` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: deny-all
  namespace: admin
spec:

>   {}

Author: I think it also works without this `{}`, I'm keeping it just to make the it more obvious that the `spec` is empty.

>   {}

Reviewer 4: I see. thanks.

> ### Enabling authorization

Reviewer 4: Just call it "Implicit Enablement", to be clear. Otherwise, people may think they need to enable the feature somehow.

There is no need to explicitly enable Istio authorization feature, you just apply
the `AuthorizationPolicy` on **workloads** to enforce the access control.

If no `AuthorizationPolicy` applies to a workload, no access control will be enforced,

> in other words, all requests will be allowed.

Reviewer 4: NIT: in other words => In other words

If any `AuthorizationPolicy` applies to a workload, the request will be allowed
only if it matches the `AuthorizationPolicy`, in other words, the request will be

> denied if it doesn't match any of the `AuthorizationPolicy`, i.e. deny by default.

Reviewer 4: If any AuthorizationPolicy applies to a workload, access to that workload is denied by default, unless explicitly allowed by a rule declared in the policy.

This is the same behavior as how [Kubernetes network policy](https://kubernetes.io/docs/concepts/services-networking/network-policies/)

> works today.

Reviewer 4: We do not need to mention it.

the request context against the current authorization policies, and returns the
authorization result, `ALLOW` or `DENY`.

### Implicit Enablement

> There is no need to explicitly enable Istio authorization feature, you just apply

Reviewer 2: There is no need to explicitly enable Istio's authorization feature, you just apply

> the `AuthorizationPolicy` on **workloads** to enforce the access control.

Reviewer 2: the `AuthorizationPolicy` on **workloads** to enforce ~the~ access control.

If no `AuthorizationPolicy` applies to a workload, no access control will be enforced,
In other words, all requests will be allowed.

If any `AuthorizationPolicy` applies to a workload, access to that workload is
denied by default, unless explicitly allowed by a rule declared in the policy.

Currently `AuthorizationPolicy` only supports "ALLOW" action. This means that if
multiple authorization policies apply to the same workload, the effect is additive.

### Authorization policy

To configure an Istio authorization policy, you specify the [`AuthorizationPolicy` configuration object](
/docs/reference/config/authorization/authorization-policy/). Like other Istio configuration objects, it

> is defined as Kubernetes `CustomResourceDefinition` [(CRD)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)

Reviewer 2: d

object.

> An `AuthorizationPolicy` includes a `selector` and a list of `rules`, the `selector`

Reviewer 2: An `AuthorizationPolicy` includes a `selector` and a list of `rules` ~,~ . ~t~ The `selector`

- **what** refers to the `to` section in the `rule` of the `AuthorizationPolicy`.
- **conditions** refers to the `when` section in the `rule` of the `AuthorizationPolicy`.

Each rule has the following standard fields:

- **`from`**: A list of sources.
- **`to`**: A list of operations.
- **`when`**: A list of custom conditions.

The following example shows an `AuthorizationPolicy` that allows two sources
(service account `cluster.local/ns/default/sa/sleep` and namespace `dev`) to access the
workloads with labels `app: httpbin` and `version: v1` in namespace foo when the request
is sent with a valid JWT token.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: httpbin
 namespace: foo
spec:
 selector:
   matchLabels:
     app: httpbin
     version: v1
 rules:
 - from:
   - source:
       principals: ["cluster.local/ns/default/sa/sleep"]
   - source:
       namespaces: ["dev"]
   to:
   - operation:
       methods: ["GET"]
   when:
   - key: request.auth.claims[iss]
     values: ["https://accounts.google.com"]
{{< /text >}}

#### Policy Target

Policy scope (target) is determined by `metadata/namespace` and an optional `selector`.

> The `metadata/namespace` tells which namespace the policy applies. If set to the

Reviewer 2: The `metadata/namespace` tells which namespace the policy applies to. If set to the

root namespace, the policy applies to all namespaces in a mesh. The value of
root namespace is configurable, and the default is `istio-system`. If set to a
normal namespace, the policy will only apply to the specified namespace.

A workload `selector` can be used to further restrict where a policy applies.
The `selector` uses pod labels to select the target workload. The workload
selector contains a list of `{key: value}` pairs, where the `key` is the name of the label.
If not set, the authorization policy will be applied to all workloads in the same namespace
as the authorization policy.

> The following example policy `allow-read` which allows `"GET"` and `"HEAD"` access to

Reviewer 2: The following example policy `allow-read` ~which~ allows `"GET"` and `"HEAD"` access to

the workload with label `app: products` in the `default` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-read
  namespace: default
spec:
  selector:
    matchLabels:
      app: products
  rules:
  - to:
    - operation:
         methods: ["GET", "HEAD"]
{{< /text >}}

#### Value matching

Exact match, prefix match, suffix match, and presence match are supported for most
of the field with a few exceptions (e.g., the `key` field under `when` section,

> the `ipBlocks` under `source` section and the `ports` field under `to` section only supports exact match).

Reviewer 2: the `ipBlocks` under `source` section and the `ports` field under `to` section only support~s~ exact match).

This means that the field can match any value, but it cannot be empty. Note that it is different from leaving a field unspecified, which means anything including empty.

The following example policy allows allows access at paths with prefix `"/test/"` or suffix `"/info"`.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: tester
  namespace: default
spec:
  selector:
    matchLabels:
      app: products
  rules:
  - to:
    - operation:
        paths: ["/test/*", "*/info"]
{{< /text >}}

#### Allow-all and deny-all

The example below shows a simple policy `allow-all` which allows full access to all
workloads in the `default` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-all
  namespace: default
spec:
  rules:
  - {}
{{< /text >}}

The example below shows a simple policy `deny-all` which denies access to all workloads
in the `admin` namespace.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: deny-all
  namespace: admin
spec:
  {}
{{< /text >}}

#### Custom conditions

> You can also use the `when` section to specify additional conditions, for example, the following

Reviewer 2: You can also use the `when` section to specify additional conditions.~,~ ~f~For example, the following

`AuthorizationPolicy` definition includes a condition that `request.headers[version]` is either `"v1"` or `"v2"`.
In this case, the key is `request.headers[version]`, which is an entry in the Istio attribute `request.headers`,
which is `map` data type.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: httpbin
 namespace: foo
spec:
 selector:
   matchLabels:
     app: httpbin
     version: v1
 rules:
 - from:
   - source:
       principals: ["cluster.local/ns/default/sa/sleep"]
   to:
   - operation:
       methods: ["GET"]
   when:
   - key: request.headers[version]
     values: ["v1", "v2"]
{{< /text >}}

The supported `key` values of a condition are listed in the
[conditions page](/docs/reference/config/authorization/conditions/).

#### Authenticated and unauthenticated identity

> In case you want to make a workload publicly accessible, you need to leave the

Reviewer 2: ~In case~ If you want to make a workload publicly accessible, you need to leave the

kind: AuthorizationPolicy
metadata:
 name: httpbin
 namespace: foo
spec:
 selector:
   matchLabels:
     app: httpbin
     version: v1
 rules:
 - to:
   - operation:
       methods: ["GET", "POST"]
{{< /text >}}

To allow only **authenticated** users, set `principal` to `"*"` instead, for example:

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: httpbin
 namespace: foo
spec:
 selector:
   matchLabels:
     app: httpbin
     version: v1
 rules:
 - from:
   - source:
       principals: ["*"]
   to:
   - operation:
       methods: ["GET", "POST"]
{{< /text >}}

### Using Istio authorization on plain TCP protocols

Istio authorization supports workloads using any plain TCP protocols, such as MongoDB. In this case,
you configure the authorization policy in the same way you did for the HTTP workloads.
The difference is that certain fields and conditions are only applicable to HTTP workloads.
These fields include:

- The `request_principals` field in the source section of the authorization policy object
- The `hosts`, `methods` and `paths` fields in the operation section of the authorization policy object

> - The supported conditions are listed in the [conditions page](/docs/reference/config/authorization/conditions/).

Reviewer 2: - The supported conditions ~are~ listed ~i~on the [conditions page](/docs/reference/config/authorization/conditions/).

If you use any HTTP only fields for a TCP workload, Istio will ignore HTTP only fields in the
authorization policy.

Assuming you have a MongoDB service on port 27017, the following example configures an authorization

> policy to only allow the `bookinfo-ratings-v2` in the Istio mesh to access the MongoDB workload.

Reviewer 2: policy to only allow the `bookinfo-ratings-v2` service in the Istio mesh to access the MongoDB workload.
